smart_open/s3.py

```diff
@@ -657,21 +681,37 @@ def seek(self, offset, whence=constants.WHENCE_START):
         :param int whence: Where the offset is from.
 
         Returns the position after seeking."""
-        # Convert relative offset to absolute, since self._raw_reader
-        # doesn't know our current position.
-        if whence == constants.WHENCE_CURRENT:
-            whence = constants.WHENCE_START
-            offset += self._current_pos
-
-        self._current_pos = self._raw_reader.seek(offset, whence)
+        if whence not in constants.WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % constants.WHENCE_CHOICES)
+        if whence == constants.WHENCE_END and offset > 0:
+            raise ValueError('offset must be <= 0 when whence == WHENCE_END, got offset: ' + offset)
+
+        if whence == constants.WHENCE_END and self._content_length is None:
+            # WHENCE_END: head request is necessary to determine file length if it's not known yet
+            #             this API call is necessary to return the absolute position as required by io.IOBase
+            response = _head(self._client, self._bucket, self._key, self._version_id)
+            self._log_retry_attempts(response)
+            self._content_length = int(response['ContentLength'])
+            self._position = self._content_length + offset
+        elif whence == constants.WHENCE_END:
+            # WHENCE_END: we already have file length, no API call needed to compute the absolute position
+            self._position = self._content_length + offset
+        else:
```

Comment:
Better to keep things simple here.

```suggestion
        elif whence == constants.WHENCE_START:
            self._position = offset
        elif whence == constants.WHENCE_CURRENT:
            self._position += offset
        else:
             raise ValueError(...)
```

You can also get rid of the whence check at the start of the function this way.